micromasters/signals.py

```diff
@@ -0,0 +1,80 @@
+"""
+Signals for the micromasters app
+"""
+
+import logging
+
+from django.db.models.signals import (
+    pre_save,
+    post_save,
+    post_delete,
+)
+from django.dispatch import receiver
+from rolepermissions.shortcuts import assign_role, remove_role
+
+from micromasters.models import MicroMastersRole
+
+
+log = logging.getLogger(__name__)
+
+
+@receiver(pre_save, sender=MicroMastersRole, dispatch_uid="save_remove_role_from_user")
+def save_remove_role_from_user(sender, instance, **kwargs):  # pylint: disable=unused-argument
+    """
+    Signal handler that happens before a role assignment is done.
+    If the the save happens for a modification, the previous role must be removed
+    if not correspondent to other programs.
+
+    Theoretically this is not necessary with the current implementation of the
+    django-role-permission library.
+    """
+    try:
+        old_instance = MicroMastersRole.objects.get(pk=instance.pk)
+    except MicroMastersRole.DoesNotExist:
+        return
+    # if the user has the same role assigned for other programs, then the role must be kept
+    # (it is NOT specific to a program at this level)
+    # Note this check is made BEFORE the save
+    if MicroMastersRole.objects.filter(role=old_instance.role).count() > 1:
+        return
+
+    log.debug(
+        'removing role % for user %s',
+        instance.role,
+        instance.user.username,
+    )
+    remove_role(instance.user)
```

Comment:
no: this is an helper function that basically removes the logical role assignment from the user